lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -1708,43 +1953,54 @@ class BuildControl : public FuncOpPartialLoweringPattern {
                  whileSchedPtr) {
         auto &whileOp = whileSchedPtr->whileOp;
 
-        /// Insert while iter arg initialization group(s). Emit a
-        /// parallel group to assign one or more registers all at once.
-        {
-          PatternRewriter::InsertionGuard g(rewriter);
-          auto parOp = rewriter.create<calyx::ParOp>(loc);
-          rewriter.setInsertionPointToStart(parOp.getBody());
-          for (auto group : whileSchedPtr->initGroups)
-            rewriter.create<calyx::EnableOp>(loc, group.getName());
-        }
-
-        auto cond = whileOp.getConditionValue();
-        auto condGroup =
-            getComponentState().getEvaluatingGroup<calyx::CombGroupOp>(cond);
-        auto symbolAttr = FlatSymbolRefAttr::get(
-            StringAttr::get(getContext(), condGroup.sym_name()));
         auto whileCtrlOp =
-            rewriter.create<calyx::WhileOp>(loc, cond, symbolAttr);
+            buildWhileCtrlOp(whileOp, whileSchedPtr->initGroups, loc, rewriter);
         rewriter.setInsertionPointToEnd(whileCtrlOp.getBody());
-        Operation *whileBodyOp;
-        if (isa<staticlogic::PipelineWhileOp>(whileOp.getOperation()))
-          whileBodyOp =
-              rewriter.create<calyx::ParOp>(whileOp.getOperation()->getLoc());
-        else
-          whileBodyOp =
-              rewriter.create<calyx::SeqOp>(whileOp.getOperation()->getLoc());
-        auto *whileBodyOpBlock = &whileBodyOp->getRegion(0).front();
+        auto whileBodyOp =
+            rewriter.create<calyx::SeqOp>(whileOp.getOperation()->getLoc());
+        auto *whileBodyOpBlock = whileBodyOp.getBody();
 
         /// Only schedule the 'after' block. The 'before' block is
         /// implicitly scheduled when evaluating the while condition.
         LogicalResult res = buildCFGControl(path, rewriter, whileBodyOpBlock,
                                             block, whileOp.getBodyBlock());
+
         // Insert loop-latch at the end of the while group
         rewriter.setInsertionPointToEnd(whileBodyOpBlock);
         rewriter.create<calyx::EnableOp>(
             loc, getComponentState().getWhileLatchGroup(whileOp).getName());
+
         if (res.failed())
           return res;
+      } else if (auto *pipeSchedPtr = std::get_if<PipelineScheduleable>(&group);
```

Comment:
Maybe this is for a follow-up commit, but can this `if... else if ...` sequence be rewritten to an `llvm::TypeSwitch`+factoring out the if-bodies into separate functions?

Reply:
I'm not sure the best way to replace this with a TypeSwitch instead of checking with std::get_if. The Schedulable variant would need to be in terms of mlir Ops, which WhileSchedulable and PipelineSchedulable are not. If Schedulable became a class, we could add a `Operation *getOperation()` method, and switch on that. Is that what you mean? Regardless, I think I will leave this as a follow-up since we are already chaining std::get_if calls here.